### PR TITLE
Add unknown parameters estimation

### PR DESCRIPTION
**Pull request by ErikQQY**
Part of #336

### CODE REVIEW

**Comment by ErikQQY** [comment]
The JET test failings in MIRK CI are weird, they are reproducable in Julia version 1.10 but not in Julia 1.11🤔

File: docs/src/tutorials/unknown_parameters.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+with boundary conditions of
+
+```math
+y_2(0)=0,\ y_2(\pi)=0
+```
+
+It is worthnoting that in this system, while we have two differetial equations, it isn't enough to estimate the unknown parameters and gurantee a unique numerical solution with only two given boundary conditions. While under the hood, the parameters are estimated simultanously with the numerical solution, it makes the boundary value problem an underconstrained BVP if the number of constraints are equal to the number of states, which may result in more than one solution. So we should provide additional constraint $y(0)=1$ to make sure unique numerical solutions and the estimated parameters are we actually wanted.
+
+With BoundaryValueDiffEq.jl, it's easy to solve boundary value problems with unknown parameters, we can just specify `fit_parameters=true` when constructing the BVP and provide the guess of the unknow parameters in `prob.p`, for example, to estimate the unknown parameters in the above BVP system:
```
Comment (ChrisRackauckas) [inline docs/src/tutorials/unknown_parameters.md:34]:
```suggestion
With BoundaryValueDiffEq.jl, it's easy to solve boundary value problems with unknown parameters, we can just specify `fit_parameters=true` when constructing the BVP and provide the guess of the unknown parameters in `prob.p`, for example, to estimate the unknown parameters in the above BVP system:
```

File: lib/BoundaryValueDiffEqFIRK/src/firk.jl [hunk 2]
```diff
@@ -240,7 +263,17 @@ function init_expanded(
     bcresid_prototype = __vec(bcresid_prototype)
     f,
     bc = if X isa AbstractVector
-        prob.f, prob.f.bc
+        if fit_parameters == true
+            l_parameters = length(prob.p)
+            vecf! = function (du, u, p, t)
+                prob.f(du, u, u[(end - l_parameters + 1):end], t)
```
Comment (ChrisRackauckas) [inline lib/BoundaryValueDiffEqFIRK/src/firk.jl:269]:
```suggestion
                prob.f(du, u, @view(u[(end - l_parameters + 1):end]), t)
```

File: lib/BoundaryValueDiffEqFIRK/src/firk.jl [hunk 3]
```diff
@@ -314,19 +350,28 @@ function SciMLBase.solve!(cache::FIRKCacheExpand{iip, T}) where {iip, T}
...
-    u = shrink_y(
-        [reshape(y, cache.in_size) for y in cache.y₀], length(cache.mesh), cache.stage)
+    # Parameter estimation, put the estimated parameters to sol.prob.p
+    if fit_parameters
+        length_u = cache.M - length(prob.p)
+        prob = remake(prob; p = first(cache.y₀)[(length_u + 1):end])
+        map(x -> resize!(x, length_u), cache.y₀)
+        resize!(cache.fᵢ₂_cache, length_u)
+    end
+
```
Comment (ChrisRackauckas) [inline lib/BoundaryValueDiffEqFIRK/src/firk.jl:360]: We'll need to support the SciMLStructures interface, but assuming `p` is a vector is fine for now.

File: lib/BoundaryValueDiffEqFIRK/src/interpolation.jl [hunk 4]
```diff
@@ -286,10 +292,10 @@ end
 
     # Load interpolation residual
     for jj in 1:stage
-        K[:, jj] = cache.y[ctr_y + jj].du
+        K[1:length_dz, jj] = cache.y[ctr_y + jj].du
     end
 
-    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, K) # Evaluate q(x) at midpoints
+    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, K[1:length_dz, :]) # Evaluate q(x) at midpoints
```
Comment (ChrisRackauckas) [inline lib/BoundaryValueDiffEqFIRK/src/interpolation.jl:298]:
```suggestion
    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, @view(K[1:length_dz, :])) # Evaluate q(x) at midpoints
```

File: lib/BoundaryValueDiffEqFIRK/src/interpolation.jl [hunk 5]
```diff
@@ -244,10 +249,10 @@ end
 
     # Load interpolation residual
     for jj in 1:stage
-        K[:, jj] = cache.y[ctr_y + jj].du
+        K[1:length_z, jj] = cache.y[ctr_y + jj].du
     end
 
-    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, K) # Evaluate q(x) at midpoints
+    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, K[1:length_z, :]) # Evaluate q(x) at midpoints
```
Comment (ChrisRackauckas) [inline lib/BoundaryValueDiffEqFIRK/src/interpolation.jl:255]:
```suggestion
    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, @view(K[1:length_z, :])) # Evaluate q(x) at midpoints
```

File: lib/BoundaryValueDiffEqFIRK/src/interpolation.jl [hunk 6]
```diff
@@ -156,13 +160,13 @@ end
...
     nestprob_p[2] = mesh_dt[j]
-    nestprob_p[3:end] .= yᵢ
+    nestprob_p[3:end] .= ifelse(fit_parameters, vcat(yᵢ, cache.p), yᵢ)
 
     _nestprob = remake(nest_prob, p = nestprob_p)
     nestsol = __solve(_nestprob, nest_nlsolve_alg; alg.nested_nlsolve_kwargs...)
     K = nestsol.u
 
-    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, K)
+    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, K[1:length_dz, :])
```
Comment (ChrisRackauckas) [inline lib/BoundaryValueDiffEqFIRK/src/interpolation.jl:169]:
```suggestion
    z₁, z₁′ = eval_q(yᵢ, 0.5, h, q_coeff, @view(K[1:length_dz, :]))
```

File: lib/BoundaryValueDiffEqMIRK/src/mirk.jl [hunk 7]
```diff
@@ -82,7 +88,18 @@ function SciMLBase.__init(
     bcresid_prototype = __vec(bcresid_prototype)
     f,
     bc = if X isa AbstractVector
-        prob.f, prob.f.bc
+        #TODO: Simplify the logic by wrapping the functions
+        if fit_parameters == true
+            l_parameters = length(prob.p)
+            vecf! = function (du, u, p, t)
+                prob.f(du, u, u[(end - l_parameters + 1):end], t)
```
Comment (ChrisRackauckas) [inline lib/BoundaryValueDiffEqMIRK/src/mirk.jl:95]:
```suggestion
                prob.f(du, u, @view(u[(end - l_parameters + 1):end]), t)
```

File: lib/BoundaryValueDiffEqFIRK/src/interpolation.jl [hunk 8]
```diff
@@ -156,13 +160,13 @@ end
 
     nestprob_p[1] = mesh[j]
     nestprob_p[2] = mesh_dt[j]
-    nestprob_p[3:end] .= yᵢ
+    nestprob_p[3:end] .= ifelse(fit_parameters, vcat(yᵢ, cache.p), yᵢ)
```
[the remark this answers is not in the document]
Comment (ErikQQY) [reply lib/BoundaryValueDiffEqFIRK/src/interpolation.jl:163]: Actually, `ifelse` should not be broadcasted here
